libraries/botbuilder-ai/src/qnaMakerDialog.ts

```diff
@@ -0,0 +1,359 @@
+            cardNoMatchText: this.cardNoMatchText,
+            noAnswer: this.noAnswer
+        };
+    }
+
+    /**
+     * Queries the knowledgebase and either passes result to the next step or constructs and displays an active learning card
+     * if active learning is enabled and multiple score close answers are returned.
+    **/
+    private async callGenerateAnswer(step: WaterfallStepContext): Promise<DialogTurnResult> {
+        const dialogOptions: QnAMakerDialogOptions = step.activeDialog.state[this.options];
+        dialogOptions.qnaMakerOptions.qnaId = 0;
+        dialogOptions.qnaMakerOptions.context = { previousQnAId: 0, previousUserQuery: '' };
+
+        step.values[this.currentQuery] = step.context.activity.text;
+        const previousContextData: { [key: string]: number } = step.activeDialog.state[this.qnAContextData] ? step.activeDialog.state[this.qnAContextData] : {};
+        var previousQnAId = step.activeDialog.state[this.previousQnAId] ? step.activeDialog.state[this.previousQnAId] : 0;
+        
+        if (previousQnAId > 0) {
+            dialogOptions.qnaMakerOptions.context = { previousQnAId: previousQnAId, previousUserQuery: '' };
+            
+            if (previousContextData[step.context.activity.text]) {
+                dialogOptions.qnaMakerOptions.qnaId = previousContextData[step.context.activity.text];
+            }
+        }
+        
+        const qna = this.getQnAClient();
+            
+        const response = await qna.getAnswersRaw(step.context, dialogOptions.qnaMakerOptions);
+            
+        const qnaResponse = {
+            activeLearningEnabled: response.activeLearningEnabled,
+            answers: response.answers
+        };
+
+        previousQnAId = -1;
+        step.activeDialog.state[this.previousQnAId] = previousQnAId;
+        const isActiveLearningEnabled = qnaResponse.activeLearningEnabled;
+        
+        step.values[this.qnAData] = response.answers;
+        
+        if (isActiveLearningEnabled && qnaResponse.answers.length > 0 && qnaResponse.answers[0].score <= this.maximumScoreForLowScoreVariation) {
+            qnaResponse.answers = qna.getLowScoreVariation(qnaResponse.answers);
+            
+            if (qnaResponse.answers && qnaResponse.answers.length > 1) {
+                var suggestedQuestions: string[] = [];
+                
+                qnaResponse.answers.forEach(answer => {
+                    suggestedQuestions.push(answer.questions[0]);
+                });
+
+                var message = new QnACardBuilder().GetSuggestionsCard(suggestedQuestions, dialogOptions.qnaDialogResponseOptions.activeLearningCardTitle, dialogOptions.qnaDialogResponseOptions.cardNoMatchText); 
+                await step.context.sendActivity(message);
+
+                step.activeDialog.state[this.options] = dialogOptions;
+            
+                return { status: DialogTurnStatus.waiting, result: undefined };
```

Comment:
>{ status: DialogTurnStatus.waiting, result: undefined } [](start = 23, length = 55)

You can just return Dialog.EndOfTurn

Reply:
This is going to cause the waterfall to skip to the next step when it continues.

---
In reply to: [383549009](https://github.com/microsoft/botbuilder-js/pull/1769#discussion_r383549009) [](ancestors = 383549009)

Reply:
Done and thanks.